### ISSUE TITLE
Refactor index handling in Hyrise

### ISSUE THREAD

**Issue by Bouncner** [hyrise/hyrise#1969]
Indexes are handled via few functions right now in a way that might cause problems for growing tables. Also, the interfaces of `Chunk` and `AbstractIndex` need some love.

A few points to tackle:
* Keep the IndexStatistics per table, but store an additional vector of ChunkIDs for the chunk being indexed
* Indexes should know about the column IDs they cover, but not the base segments.
* Move the IndexScan operators into the table scan by adding a new impl.

Did I forget something?

**Comment by mrks** [comment]
Here's what I remember from the discussion:

> Keep the IndexStatistics per table, but store an additional vector of ChunkIDs for the chunk being indexed

Also, remove the information about the index type.

> Indexes should know about the column IDs they cover, but not the base segments.

No, the chunk should have a `column ID -> index` mapping; the base indexes should not know about the segments. For some indexes, e.g., those that use value ids, a pointer to either the dictionary or the segment is fine; but only for internal purposes, not for establishing an index-segment relationship.

> Move the IndexScan operators into the table scan by adding a new impl.

This is #1790.

**Comment by Bouncner** [comment]
I started implementing it a while ago and haven't had the time to finish the branch. Since there will be some work on this part in the near future, I wanted to bring up an issue for what I do not have a good idea.

The problem is the following: let's assume we have an index on columns 3,2,1 (column 3 is the most frequently accessed, thus the order). Now, we have a predicate chain of 1,2,3 which would be answered by the index.
How exactly do we want to implement this? I thought about sets, other data structures, and `std::search`, but I did not come up with a solution that is really better than iterating over all possible combinations (an index 3,2,1 could perfectly answer two predicates of 4,2,3).
Further, some indexes allow certain predicates, others don't (e.g., range predicates in group keys vs. in ART index). Do we want the flexibility to let the index determine what it can answer or will there a central place to do that?

**Comment by mrks** [comment]
This sounds like a problem that we might look at some time later. In general, the optimizer should make decisions that go beyond the internal execution strategy for a single operator. Potentially reordering predicate chains to exploit indexes certainly sounds like optimizer territory.

Once that decision has been made, it should be as simple as passing a conjunction expression into the TableScan. `TableScan::create_impl` could then produce a combined index scan impl or multiple independent impls (for unindexed segments).

To me, this sounds that most of this could be contained in `TableScan::_on_execute` and `create_impl` with only minor changes necessary to the callers and the impls themselves. Potentially, this could even have benefits for conjunctive scans on reference segments, as we would only have to do the relative expensive step of materializing the output PosList (1) once and (2) only for the final, smallest output:

https://github.com/hyrise/hyrise/blob/master/src/lib/operators/table_scan.cpp#L120-L153

These are just first thoughts, don't mistake it for a building plan ;)

**Comment by Bouncner** [comment]
Sounds good to me, but we should not underestimate the handling of predicate chains and the various characteristics of our indexes. I don't have a problem with dropping the B-Tree or ART indexes in the long run, but that's another discussion.

So what would be an acceptable interface to the chunk? Just a call to get a vector of all indexes? A implemented a `applicable_indexes` but it becomes immediately hard to tell where to step (set intersections? Does need to fully cover the requested columns? Care about the order?).

**Comment by mrks** [comment]
Do we have a list of all potential predicates that should be covered by indexes and which index supports them? This seems like something the optimizer should know.

**Comment by Bouncner** [comment]
Not that I am aware of. But definitely something we'll need.

So can we agree on that minimalistic chunk-index interface?

**Comment by mrks** [comment]
At least I can't see anything wrong with it right now.

**Comment by Bensk1** [comment]
> Not that I am aware of. But definitely something we'll need.
> 
> So can we agree on that minimalistic chunk-index interface?

Yes, I also believe that the optimizer should take the decisions and determine whether an index is applicable. I think @mweisgut has #1790 already on his radar.

**Comment by mrks** [comment]
@julianmenzler will also look into index questions, but from the perspective of transactional workloads. In the TPC-C, the same data is updated over and over again, so we end up with massive tables that hold only tens of visible rows. An index is a first step, but would still return thousands of rows. By iterating over the results from the end and validating them immediately, we can use an efficient early out. Let's include him in these discussions.

**Comment by mrks** [comment]
Notes from our meeting:
- Indexes are currently not used by default
- New index-based join is currently only used when PQPs are built by hand; index-based scan is theoretically implemented, but the optimizer does not use them for TPC-H.
- Is TableScanImpl the best solution or do we want something similar to `ColumnVsValueTableScanImpl::_scan_indexes_segment` better? Pro: Get ReferenceSegment handling for free
- Something to consider: If we want "enough already" handling (i.e., searching from the backing, stopping at the first valid row), the indexed and non-indexed code need a way to communicate.
- Open question: How do we handle multi-column indexes? TableScan could accept conjunctive expressions.
- After the refactoring, we should benchmark the indexes again.
- Table::IndexStatistics should stay. We might want to keep the number of indexed chunks, but for now, we simply assume that all chunks are indexed.
- Differing from #1790, the optimizer should not pass a list of chunks that should be scanned using an index.
- For each chunk, the index to use can be identified by the list of queried attributes. Keep in mind that the index a,b,c could be used for a predicate on a and b. We do not consider or anticipate cases where different types (group key, btree) might exist on the same data.
- One advantage: If the TableScan decides which indexes to use, we will not ignore new indexes that were created since the PQP was built/cached.
- AbstractIndex keeps indexed column IDs, not pointers to the segments (reconfirmed from above).
- We do not want to consider multi-chunk indexes
- We should check if multi-column indexes for `NULL` are necessary. There is a tendency to ignore these and close #1818. Multi-column indexes ignore NULL values and those have to be queried using regular scans.
- We are not looking into enforcing unique constraints for now.
- When benchmarking the index-based join, we should test cases where the last chunks are not indexed. If the nested loop join breaks our neck, we might need the inclusion/exclusion approach again.

Next steps:
- @julianmenzler will look at the index scan refactoring
- @mweisgut will finish IndexJoin, integrate it so that it is actually used.

**Comment by mweisgut** [comment]
>We are not looking into enforcing unique constraints for now.

What did we mean by that?

**Comment by mrks** [comment]
Our constraints (PRIMARY, UNIQUE) are currently only modeled as "soft constraints". This means that we simply promise to the optimizer that these constraints are true. If you were to insert duplicate values, nobody would complain and you might end up with wrong results.

**Comment by mweisgut** [comment]
>We do not want to consider multi-chunk indexes

Recent experiments showed that calling `_index_ranges_for_value` takes most of the execution time of the overall IndexJoin execution. Within this function, `upper_bound` / `lower_bound` of the existing indexes are called which in turn call `upper_bound` / `lower_bound` of the underlying dictionary segment. These calls take up a significant part of the IndexJoin's execution time. The higher the number of indexes, the higher the number of dictionary lookups and the slower the IndexJoin.
Consequently, as discussed in person recently, multi-chunk indices should be considered.

>When benchmarking the index-based join, we should test cases where the last chunks are not indexed. If the nested loop join breaks our neck, we might need the inclusion/exclusion approach again.

For each below-listed commit, TPC-H  Q19 was benchmarked with the following command on sidon, pinned to NUMA node 2:
`./hyriseBenchmarkTPCH -q 19 -t 20 -s 1 -o q19_<commit hash>.b --dont_cache_binary_tables --indexes`

- 342837c88: index join used at join position 1 (see PQP below), all lineitem chunks indexed
- b703680a3: index join used at join position 1 (see PQP below), all lineitem chunks except the last one indexed (nested loop fallback join execution)

**all lineitem chunks indexed v.s. all lineitem chunks except the last one indexed**
```
../scripts/compare_benchmarks.py q19_342837c88.b q19_b703680a3.b  
+----------------+-------------------+------+-----------------+------+------------+----------------------+ 
| Benchmark      | prev. iter/s      | runs | new iter/s      | runs | change [%] |              p-value | 
+----------------+-------------------+------+-----------------+------+------------+----------------------+ 
| TPC-H 19       | 20.22556495666504 | 405  | 4.4613938331604 | 90   | -78%       | (run time too short) | 
| geometric mean |                   |      |                 |      | -78%       |                      | 
+----------------+-------------------+------+-----------------+------+------------+----------------------+
```

**TPC-H Q19 PQP**
![TPC-H_19-PQP_hash_joins_enumerated](https://user-images.githubusercontent.com/20044656/82899722-bd994400-9f5b-11ea-955d-3c59f64b064c.png)

>@mweisgut will finish IndexJoin, integrate it so that it is actually used.

After #1851 and #2121, the IndexJoin can be considered as "finished" for now. However, with the current index-per-chunk approach, the performance of an IndexJoin depends highly on the number of chunks (see first paragraph). Further knowledge about the influence of different chunk sizes / numbers of chunks is required in order to (1) adapt our indexing approach (2) integrate the IndexJoin so that it is actually used.
